src/enzyme_ad/jax/Utils.cpp

```diff
@@ -566,6 +574,183 @@ bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy,
   return allowOnFloatingPointMath;
 }
 
+bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy, Type inTy,
+                          mlir::Operation *op) {
+  outTy = getElementTypeOrSelf(outTy);
+  inTy = getElementTypeOrSelf(inTy);
+  if (outTy.isInteger() && inTy.isInteger())
+    return true;
+  return allowOnFloatingPointMath || guaranteedNoNanResult(op);
+}
+
+bool guaranteedNoNanResult(stablehlo::ConstantOp constOp) {
+  Attribute attr = constOp.getValue();
+
+  if (auto denseAttr = dyn_cast<DenseElementsAttr>(attr)) {
+    if (denseAttr.getType().getShape().size() && denseAttr.isSplat()) {
+      denseAttr = denseAttr.resizeSplat(
+          RankedTensorType::get({}, denseAttr.getType().getElementType()));
+    }
+    // For floating point values
+    if (isa<FloatType>(denseAttr.getElementType())) {
+      for (auto element : denseAttr.getValues<APFloat>()) {
+        if (element.isNaN())
+          return false;
+      }
+      return true;
+    }
+
+    if (denseAttr.getElementType().isIntOrIndex()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool guaranteedNoNanResult(mlir::Value value) {
+  return guaranteedNoNanResult(value.getDefiningOp());
+}
+
+bool guaranteedNoNanResult(mlir::Operation *op) {
+  if (!op)
+    return false;
+
+  if (auto constantOp = dyn_cast<mlir::stablehlo::ConstantOp>(op)) {
+    return guaranteedNoNanResult(constantOp);
+  }
+
+  if (isa<stablehlo::AbsOp, stablehlo::ExpOp, stablehlo::TanhOp,
+          stablehlo::AndOp, stablehlo::OrOp, stablehlo::XorOp, stablehlo::NotOp,
+          stablehlo::AddOp, stablehlo::SubtractOp, stablehlo::MulOp,
+          stablehlo::SineOp, stablehlo::CosineOp, stablehlo::ConvertOp,
+          stablehlo::SliceOp, stablehlo::ConcatenateOp,
+          stablehlo::BroadcastInDimOp>(op)) {
+    for (auto operand : op->getOperands()) {
+      if (!guaranteedNoNanResult(operand)) {
+        return false;
```

Comment:
this isn't sufficient atm, 0 and inf are non nan, but multiplying gives nan [same with inf - inf, sin(inf)].

data movement ops [slice/concat/bcast/reshape/transpose/etc] this is valid for. integer ones [and/or/xor/nan] probably can be ignored as already handled per integer?. convert/abs I think this also applies to. others I think may need some more thought